gemrb/core/Map.cpp

```diff
@@ -2793,246 +2796,203 @@ PathNode* Map::GetLine(const Point &start, const Point &dest, int Speed, int Ori
-		Leveldown( p.x - 1, p.y + 1, level, n, diff );
-		Leveldown( p.x + 1, p.y + 1, level, n, diff );
-		Leveldown( p.x + 1, p.y - 1, level, n, diff );
-		Leveldown( p.x - 1, p.y - 1, level, n, diff );
-		if (!diff)
-			return Return;
-
-		if (fixup_orient) {
-			// don't change orientation at end of path? this seems best
-			StartNode->orient = GetOrient( p, n );
-		}
-
-		Return = new PathNode;
-		Return->Next = StartNode;
-		Return->Next->Parent = Return;
-		StartNode = Return;
-
-		StartNode->x = n.x;
-		StartNode->y = n.y;
-		StartNode->orient = GetOrient( p, n );
-		p = n;
-	}
-
-	return Return;
-}
-
-PathNode* Map::FindPath(const Point &s, const Point &d, unsigned int size, int MinDistance)
-{
-	Point start( s.x/16, s.y/12 );
-	Point goal ( d.x/16, d.y/12 );
-	memset( MapSet, 0, Width * Height * sizeof( unsigned short ) );
-	while (InternalStack.size())
-		InternalStack.pop();
-
-	if (GetBlocked( d.x, d.y, size )) {
-		AdjustPosition( goal );
-	}
-	unsigned int pos = ( goal.x << 16 ) | goal.y;
-	unsigned int pos2 = ( start.x << 16 ) | start.y;
-	InternalStack.push( pos );
-	MapSet[goal.y * Width + goal.x] = 1;
-
-	while (InternalStack.size()) {
-		pos = InternalStack.front();
-		InternalStack.pop();
-		unsigned int x = pos >> 16;
-		unsigned int y = pos & 0xffff;
-
-		if (pos == pos2) {
-			//We've found _a_ path
-			//print("GOAL!!!");
-			break;
+		// Lazy Theta-star optimization: does the LOS check after inserting a node
+		smptParent = parents[smptCurrent.y * Width + smptCurrent.x];
+		nmptParent.x = smptParent.x * 16;
+		nmptParent.y = smptParent.y * 12;
+		bool reachable = IsVisibleLOS(nmptCurrent, nmptParent, true);
+		if (!reachable && smptCurrent != smptSource) {
+			smptBestParent.x = 0;
+			smptBestParent.y = 0;
+			distFromStart[smptCurrent.y * Width + smptCurrent.x] = MAX_PATH_COST;
+			for (size_t i = 0; i < DEGREES_OF_FREEDOM; i++) {
+				smptCurNeighbor.x = smptCurrent.x + dx[i];
+				smptCurNeighbor.y = smptCurrent.y + dy[i];
+				if (isClosed[smptCurNeighbor.y * Width + smptCurNeighbor.x]) {
+					unsigned short curParentDist = distFromStart[smptCurNeighbor.y * Width + smptCurNeighbor.x] +
+									Distance(smptCurrent, smptCurNeighbor);
+					if (curParentDist < distFromStart[smptCurrent.y * Width + smptCurrent.x]) {
+						smptBestParent = smptCurNeighbor;
+						distFromStart[smptCurrent.y * Width + smptCurrent.x] = curParentDist;
+					}
+				}
+				parents[smptCurrent.y * Width + smptCurrent.x] = smptBestParent;
+			}
 		}
-		unsigned int Cost = MapSet[y * Width + x] + NormalCost;
-		if (Cost > 65500) {
-			//print("Path not found!");
-			break;
+		isClosed[smptCurrent.y * Width + smptCurrent.x] = true;
+		for (size_t i = 0; i < DEGREES_OF_FREEDOM; i++) {
+			smptChild.x = smptCurrent.x + dx[i];
+			smptChild.y = smptCurrent.y + dy[i];
+			bool childOutsideMap = smptChild.x < 0 ||
+							  smptChild.y < 0 ||
+							  (unsigned) smptChild.x >= Width ||
+							  (unsigned) smptChild.y >= Height;
```

Comment:
There appear to be extra spaces here for indentation. Should only be tabs and I think it'd also look better without an extra level of indentation, since all the statements are of the same level.

And apparently we lost our draft coding style, since titanpad went down and the wayback machine didn't manage to get a good snapshot. Grrr.